- Implemented delete/move events.

vicng opened the pull request:
- Changed undo of CreateEvent to delete instead of move to trash

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-ios/323)
<!-- Reviewable:end -->

vicng commented:
+@corydcode +@roboerikg 

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/323#-:-KeaNND9j0BF30Lv217D:b4ivu6b)*
<!-- Sent from Reviewable.io -->

RoboErikG commented on Source/Layout/BlockGroupLayout.swift:
```diff
@@ -201,8 +201,19 @@ open class BlockGroupLayout: Layout {
+        updatePosition()
+        try? event.recordNewValues(fromBlock: block)
+        EventManager.sharedInstance.addPendingEvent(event)
+      } else {
```
What does it mean to move a block group without any blocks in it?

RoboErikG commented on Source/UI/View Controllers/WorkbenchViewController.swift:
```diff
@@ -917,15 +929,21 @@ extension WorkbenchViewController: EventManagerListener {
-    EventManager.sharedInstance.isEnabled = true
+    // Fire pending events before listening to events again, in case outside listeners need to
+    // update their state from those events.
+    EventManager.sharedInstance.firePendingEvents()
```
Is the end of PanGestureRecognizer the only place that calls firePendingEvents() other than undo/redo? Who fires events if a user deletes a variable and all associated blocks? Or if the developer mutates a block in response to another event?

RoboErikG commented on Source/UI/View Controllers/WorkbenchViewController.swift:
```diff
@@ -972,25 +999,117 @@ extension WorkbenchViewController {
+
+    // Check current parent of block
+    if let inferiorConnection = block.inferiorConnection {
+      if let currentParent = inferiorConnection.targetBlock, currentParent.uuid == parentID {
```
Might be nice to know how you get here. Seems like this would be a bug.

RoboErikG commented on Source/UI/View Controllers/WorkbenchViewController.swift:
```diff
@@ -972,25 +999,117 @@ extension WorkbenchViewController {
+      // Move to new workspace position
+      blockLayout.rootBlockGroupLayout?.move(toWorkspacePosition: position)
+    } else if let inferiorConnection = block.inferiorConnection,
+      let parentID = parentID,
```
Does this need to be let aParentId = parentID like above?

RoboErikG commented on Source/UI/View Controllers/WorkbenchViewController.swift:
```diff
@@ -972,25 +999,117 @@ extension WorkbenchViewController {
+   - parameter runForward: Flag determining if the event should be run forward (`true` for redo
+   operations) or run backward (`false` for undo operations).
+   */
+  open func update(fromMoveEvent event: MoveEvent, runForward: Bool) {
```
There are a lot of errors that indicate we're in a pretty bad state that are being masked by those debug prints. How do we get into a bad state and is it actually recoverable?

vicng commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/Layout/BlockGroupLayout.swift, line 214 at r1](https://reviewable.io:443/reviews/google/blockly-ios/323#-KeaNLJaIL5PUmB9IseV-r1-214:-KeeXiSRFo8tNdGq1HgV:bdtdo4q) ([raw file](https://github.com/google/blockly-ios/blob/ef9c4d5ab8934cac6b776890a9b0fd878a38163a/Source/Layout/BlockGroupLayout.swift#L214)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

What does it mean to move a block group without any blocks in it?
</blockquote></details>

For events, nothing.

---

*[Source/UI/View Controllers/WorkbenchViewController.swift, line 934 at r1](https://reviewable.io:443/reviews/google/blockly-ios/323#-KeaNLJboIJv-uPeHRGV-r1-934:-KeeXt8tdn37dpvPf8go:bvwa77k) ([raw file](https://github.com/google/blockly-ios/blob/ef9c4d5ab8934cac6b776890a9b0fd878a38163a/Source/UI/View Controllers/WorkbenchViewController.swift#L934)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Is the end of PanGestureRecognizer the only place that calls firePendingEvents() other than undo/redo? Who fires events if a user deletes a variable and all associated blocks? Or if the developer mutates a block in response to another event?
</blockquote></details>

Right now, there are the only two places where it's being fired.

I was planning on figuring out fire locations/grouping near the end, or at least until change events are in (next CL).

---

*[Source/UI/View Controllers/WorkbenchViewController.swift, line 1055 at r1](https://reviewable.io:443/reviews/google/blockly-ios/323#-KeaNLJboIJv-uPeHRGV-r1-1055:-KefOF1i41qLDwwEo4EQ:b-31mliq) ([raw file](https://github.com/google/blockly-ios/blob/ef9c4d5ab8934cac6b776890a9b0fd878a38163a/Source/UI/View Controllers/WorkbenchViewController.swift#L1055)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

There are a lot of errors that indicate we're in a pretty bad state that are being masked by those debug prints. How do we get into a bad state and is it actually recoverable?
</blockquote></details>

Depends on the context. For real-time, it would probably lead to an unrecoverable state. For local events however, I don't *think* so. If you can't perform the operation, the app would appear buggy, but the system wouldn't be "broken" --  it just wouldn't be replaying events properly. For reference, this follows web's impl of event undo/redo.

I've changed all these calls to be bky_assertionFailure's so it fails in debug mode, but so they run fine in release. IMO, it's not an error worth crashing the app over (until we are really supporting real-time events).

---

*[Source/UI/View Controllers/WorkbenchViewController.swift, line 1075 at r1](https://reviewable.io:443/reviews/google/blockly-ios/323#-KeaNLJboIJv-uPeHRGV-r1-1075:-KefPfoW4VjXnG1ZAMtK:b-mahjlt) ([raw file](https://github.com/google/blockly-ios/blob/ef9c4d5ab8934cac6b776890a9b0fd878a38163a/Source/UI/View Controllers/WorkbenchViewController.swift#L1075)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Might be nice to know how you get here. Seems like this would be a bug.
</blockquote></details>

Perhaps there's a duplicate move event that could have fired (maybe from a real-time source). I don't think it's necessarily an error. 

---

*[Source/UI/View Controllers/WorkbenchViewController.swift, line 1090 at r1](https://reviewable.io:443/reviews/google/blockly-ios/323#-KeaNLJboIJv-uPeHRGV-r1-1090:-KefRG2_fy6UWYVmhIDs:b-2rlth0) ([raw file](https://github.com/google/blockly-ios/blob/ef9c4d5ab8934cac6b776890a9b0fd878a38163a/Source/UI/View Controllers/WorkbenchViewController.swift#L1090)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Does this need to be let aParentId = parentID like above?
</blockquote></details>

No. I thought it did initially, but you can re-scope the same variable from optional to non-optional (which I find reads better).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/323)*
<!-- Sent from Reviewable.io -->

CoryDCode commented:
Review status: 0 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Layout/WorkspaceLayoutCoordinator.swift, line 267 at r1](https://reviewable.io:443/reviews/google/blockly-ios/323#-Kee8-fvjP179F2xjhZK:-Kee8-fvjP179F2xjhZL:bws6p5w) ([raw file](https://github.com/google/blockly-ios/blob/ef9c4d5ab8934cac6b776890a9b0fd878a38163a/Source/Layout/WorkspaceLayoutCoordinator.swift#L267)):*
> ```Swift
>       EventManager.sharedInstance.addPendingEvent(event)
>     } else {
>       performDisconnect()
> ```

What is this doing? As far as I can glean, we get here if the target connection (or its source block) is nil. What are we disconnecting? (And if we ARE disconnecting something, why aren't we tracking it?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/323)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Layout/BlockGroupLayout.swift, line 214 at r1](https://reviewable.io:443/reviews/google/blockly-ios/323#-KeaNLJaIL5PUmB9IseV-r1-214:-Kej5KP9i_q6SjMlGTGK:b-zf2pxm) ([raw file](https://github.com/google/blockly-ios/blob/ef9c4d5ab8934cac6b776890a9b0fd878a38163a/Source/Layout/BlockGroupLayout.swift#L214)):*
<details><summary><i>Previously, vicng wrote…</i></summary><blockquote>

For events, nothing.
</blockquote></details>

But how do we get in this code path? When do we move an empty block group?

---

*[Source/UI/View Controllers/WorkbenchViewController.swift, line 934 at r1](https://reviewable.io:443/reviews/google/blockly-ios/323#-KeaNLJboIJv-uPeHRGV-r1-934:-Kej5XSSRT9DYIUWA0gb:bk3c5f2) ([raw file](https://github.com/google/blockly-ios/blob/ef9c4d5ab8934cac6b776890a9b0fd878a38163a/Source/UI/View Controllers/WorkbenchViewController.swift#L934)):*
<details><summary><i>Previously, vicng wrote…</i></summary><blockquote>

Right now, there are the only two places where it's being fired.

I was planning on figuring out fire locations/grouping near the end, or at least until change events are in (next CL).
</blockquote></details>

Fine to work through in another CL, but please write down somewhere your expectations for event firing and if it will explicitly be in the API or a side effect of public method calls. Developers may be calling mutate, move, or other public methods and if those don't fire events we could end up in a bad state.

---

*[Source/UI/View Controllers/WorkbenchViewController.swift, line 1075 at r1](https://reviewable.io:443/reviews/google/blockly-ios/323#-KeaNLJboIJv-uPeHRGV-r1-1075:-Kej66h0OtEu3FLHIP4G:bdcv6l8) ([raw file](https://github.com/google/blockly-ios/blob/ef9c4d5ab8934cac6b776890a9b0fd878a38163a/Source/UI/View Controllers/WorkbenchViewController.swift#L1075)):*
<details><summary><i>Previously, vicng wrote…</i></summary><blockquote>

Perhaps there's a duplicate move event that could have fired (maybe from a real-time source). I don't think it's necessarily an error. 
</blockquote></details>

Good point.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/323)*
<!-- Sent from Reviewable.io -->

vicng commented:
On second thought, I changed bky_assertionFailure back to bky_debugPrint. The library will crash in debug mode (for example, when rolling back connect events) since change events haven't been implemented in this CL.

I may change it back to bky_assertionFailure once all events are fully implemented.

---

Review status: 0 of 15 files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Layout/BlockGroupLayout.swift, line 214 at r1](https://reviewable.io:443/reviews/google/blockly-ios/323#-KeaNLJaIL5PUmB9IseV-r1-214:-KejQ1d3SzhNBDlVzm4q:btdjyzn) ([raw file](https://github.com/google/blockly-ios/blob/ef9c4d5ab8934cac6b776890a9b0fd878a38163a/Source/Layout/BlockGroupLayout.swift#L214)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

But how do we get in this code path? When do we move an empty block group?
</blockquote></details>

At the moment, there's nothing that would move an empty block group. I don't see a reason for this to change, but I don't think there's anything "wrong" with moving an empty block group.

---

*[Source/Layout/WorkspaceLayoutCoordinator.swift, line 267 at r1](https://reviewable.io:443/reviews/google/blockly-ios/323#-Kee8-fvjP179F2xjhZK:-KekQ_6QkJ7pflS7Ge5H:b-epdvet) ([raw file](https://github.com/google/blockly-ios/blob/ef9c4d5ab8934cac6b776890a9b0fd878a38163a/Source/Layout/WorkspaceLayoutCoordinator.swift#L267)):*
<details><summary><i>Previously, CoryDCode wrote…</i></summary><blockquote>

What is this doing? As far as I can glean, we get here if the target connection (or its source block) is nil. What are we disconnecting? (And if we ARE disconnecting something, why aren't we tracking it?)
</blockquote></details>

Changed to address concerns.

---

*[Source/UI/View Controllers/WorkbenchViewController.swift, line 934 at r1](https://reviewable.io:443/reviews/google/blockly-ios/323#-KeaNLJboIJv-uPeHRGV-r1-934:-KejREg4b6uekP4Ri975:bim2pwq) ([raw file](https://github.com/google/blockly-ios/blob/ef9c4d5ab8934cac6b776890a9b0fd878a38163a/Source/UI/View Controllers/WorkbenchViewController.swift#L934)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Fine to work through in another CL, but please write down somewhere your expectations for event firing and if it will explicitly be in the API or a side effect of public method calls. Developers may be calling mutate, move, or other public methods and if those don't fire events we could end up in a bad state.
</blockquote></details>

Ok, will do.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/323#-:-KekTw0aRBS-ukJlEWkV:buzb6nm)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
Couple comments left, otherwise <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Review status: 0 of 15 files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/Layout/BlockGroupLayout.swift, line 214 at r1](https://reviewable.io:443/reviews/google/blockly-ios/323#-KeaNLJaIL5PUmB9IseV-r1-214:-Keo4F2s7cuo0K6s9-39:b-xiz8ai) ([raw file](https://github.com/google/blockly-ios/blob/ef9c4d5ab8934cac6b776890a9b0fd878a38163a/Source/Layout/BlockGroupLayout.swift#L214)):*
<details><summary><i>Previously, vicng wrote…</i></summary><blockquote>

At the moment, there's nothing that would move an empty block group. I don't see a reason for this to change, but I don't think there's anything "wrong" with moving an empty block group.
</blockquote></details>

Please debug log it. This really seems like an unexpected state that could mean something is out of sync.

---

*[Source/UI/View Controllers/WorkbenchViewController.swift, line 1083 at r2](https://reviewable.io:443/reviews/google/blockly-ios/323#-Keo4uWx-iHYv45WPo8H:-Keo4uWx-iHYv45WPo8I:b-upj5lr) ([raw file](https://github.com/google/blockly-ios/blob/7bda11e379e3e7e6b60c41899807f778e1db0a97/Source/UI/View Controllers/WorkbenchViewController.swift#L1083)):*
> ```Swift
>           try _workspaceLayoutCoordinator?.disconnect(inferiorConnection)
>         } catch let error {
>           bky_debugPrint("Could not disconnect block from its parent: \(error)")
> ```

Could you start commenting why you use debug or assert for these? I don't understand the context for each case well enough to know why you're using one or the other.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/323#-:-Keo5a0sM6_tK2ZZz_QN:b-j8e322)*
<!-- Sent from Reviewable.io -->

vicng commented:
Review status: 0 of 16 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[Source/Layout/BlockGroupLayout.swift, line 214 at r1](https://reviewable.io:443/reviews/google/blockly-ios/323#-KeaNLJaIL5PUmB9IseV-r1-214:-KeoeQKiIFWTGFFIMzer:b-896fix) ([raw file](https://github.com/google/blockly-ios/blob/ef9c4d5ab8934cac6b776890a9b0fd878a38163a/Source/Layout/BlockGroupLayout.swift#L214)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Please debug log it. This really seems like an unexpected state that could mean something is out of sync.
</blockquote></details>

Done.

---

*[Source/UI/View Controllers/WorkbenchViewController.swift, line 1083 at r2](https://reviewable.io:443/reviews/google/blockly-ios/323#-Keo4uWx-iHYv45WPo8H:-KeofZDUqnqElt2jOMcX:b-896fix) ([raw file](https://github.com/google/blockly-ios/blob/7bda11e379e3e7e6b60c41899807f778e1db0a97/Source/UI/View Controllers/WorkbenchViewController.swift#L1083)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Could you start commenting why you use debug or assert for these? I don't understand the context for each case well enough to know why you're using one or the other.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/323)*
<!-- Sent from Reviewable.io -->
